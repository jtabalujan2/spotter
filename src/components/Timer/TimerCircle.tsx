import { PropsWithChildren } from "react";

interface TimerCircleProps {
   radius: number;
   circumference: number;
   height: number;
   width: number;
   timeLeft: number;
   totalTime: number;
}

export const TimerCircle = (props: PropsWithChildren<TimerCircleProps>) => {
   const { radius, circumference, height, width, timeLeft, totalTime, children } = props;

   // Calculates the amount of circle to show based on the time left
   const percentage = (timeLeft / totalTime) * 100;

   const circlePosition = (percentage / 100) * 2 * Math.PI;

   // Calculate x and y position for small circle based on the timer's radius and size (100x100)
   const circleX = 100 + radius * Math.cos(Math.PI / 2 - circlePosition);
   const circleY = 100 - radius * Math.sin(Math.PI / 2 - circlePosition);

   return (
      <svg className="w-full h-full transform -rotate-90" viewBox="0 0 200 200" role="img" aria-label="Timer progress">
         <circle
            className="text-secondary"
            strokeWidth="8"
            stroke="currentColor"
            fill="transparent"
            r={radius}
            cx={width}
            cy={height}
         />
         <circle
            className="text-white transition-all duration-300"
            strokeWidth="8"
            strokeDasharray={circumference}
            strokeDashoffset={circumference - (circumference * percentage) / 100}
            strokeLinecap="round"
            stroke="currentColor"
            fill="transparent"
            r={radius}
            cx={width}
            cy={height}
         />
         <circle
            className="text-white transition-all duration-300"
            r={12}
            cx={circleX}
            cy={circleY}
            fill="currentColor"
            transform="rotate(90 100 100)"
         />
         {children}
      </svg>
   )
}