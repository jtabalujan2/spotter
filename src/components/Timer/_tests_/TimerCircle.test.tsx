import { render } from "@testing-library/react";
import { describe, it, expect } from "vitest";

import { TimerCircle } from "../TimerCircle";

describe("TimerCircle", () => {


   it("calculates the correct percentage and circle position", () => {
      const { container } = render(
         <TimerCircle
            radius={50}
            circumference={314}
            height={100}
            width={100}
            timeLeft={25}
            totalTime={100}
         />
      );
      const circle = container.querySelector("circle:nth-of-type(2)");
      expect(circle?.getAttribute("stroke-dashoffset")).toBeCloseTo(235.5, 1);
   });

   it("renders children correctly", () => {
      const { getByText } = render(
         <TimerCircle
            radius={50}
            circumference={314}
            height={100}
            width={100}
            timeLeft={50}
            totalTime={100}
         >
            <text>Time Left</text>
         </TimerCircle>
      );
      expect(getByText("Time Left")).toBeInTheDocument();
   });

   it("renders small circle at correct position", () => {
      const { container } = render(
         <TimerCircle
            radius={50}
            circumference={314}
            height={100}
            width={100}
            timeLeft={50}
            totalTime={100}
         />
      );
      const smallCircle = container.querySelector("circle:nth-of-type(3)");
      expect(smallCircle?.getAttribute("cx")).toBeCloseTo(100, 1);
      expect(smallCircle?.getAttribute("cy")).toBeCloseTo(150, 1);
   });
});