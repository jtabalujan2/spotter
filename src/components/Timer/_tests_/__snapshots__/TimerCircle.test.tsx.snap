// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TimerCircle > renders correctly 1`] = `
<div>
  <svg
    aria-label="Timer progress"
    class="w-full h-full transform -rotate-90"
    role="img"
    viewBox="0 0 200 200"
  >
    <circle
      class="text-secondary"
      cx="100"
      cy="100"
      fill="transparent"
      r="50"
      stroke="currentColor"
      stroke-width="8"
    />
    <circle
      class="text-white transition-all duration-300"
      cx="100"
      cy="100"
      fill="transparent"
      r="50"
      stroke="currentColor"
      stroke-dasharray="314"
      stroke-dashoffset="157"
      stroke-linecap="round"
      stroke-width="8"
    />
    <circle
      class="text-white transition-all duration-300"
      cx="100"
      cy="150"
      fill="currentColor"
      r="12"
      transform="rotate(90 100 100)"
    />
  </svg>
</div>
`;

exports[`TimerCircle > renders correctly with given props 1`] = `
<div>
  <svg
    aria-label="Timer progress"
    class="w-full h-full transform -rotate-90"
    role="img"
    viewBox="0 0 200 200"
  >
    <circle
      class="text-secondary"
      cx="100"
      cy="100"
      fill="transparent"
      r="50"
      stroke="currentColor"
      stroke-width="8"
    />
    <circle
      class="text-white transition-all duration-300"
      cx="100"
      cy="100"
      fill="transparent"
      r="50"
      stroke="currentColor"
      stroke-dasharray="314"
      stroke-dashoffset="157"
      stroke-linecap="round"
      stroke-width="8"
    />
    <circle
      class="text-white transition-all duration-300"
      cx="100"
      cy="150"
      fill="currentColor"
      r="12"
      transform="rotate(90 100 100)"
    />
  </svg>
</div>
`;
